wgsl/index.bs

```diff
@@ -896,33 +900,143 @@ A <dfn noexport>type rule</dfn> has two parts:
+For example, the integer negation rule has four overloads,
+because there are four possible ways to assign a type to its type parameter `T`.
+
+Note:
+In other words, a parameterized type rule provides a pattern for a collection of concrete type rules,
+each one produced by applying a different substitution to the parameterized rule.
 
 A <dfn noexport>type rule applies to an expression</dfn> when:
 * The rule's conclusion matches a valid parse of the expression, and
 * The rule's preconditions are satisfied.
 
+A parameterized type rule applies to an expression
+if there exists a substitution producing a concrete type rule
+that applies to the expression.
+
+When a concrete type rule applies to an expression,
+we say that we have <dfn noexport>derived</dfn> the type assertion
+that is the rule's conclusion.
+
 TODO: write an example such as `1+2`, or `3 - a`, where `a` is in-scope of a let declaration with `i32` type.
 
-The type rules are designed so that if parsing succeeds, at most one type rule will apply to each expression.
-If a type rule applies to an expression, then the conclusion is asserted, and therefore determines the static type of the expression.
+We may be able to derive more than one type assertion for an expression.
+For example, the type rule for unsuffixed integer literals
+has two overloads that apply to the expression `7`:
+
+<table class='data'>
+  <thead>
+    <tr><th>Precondition<th>Conclusion
+  </thead>
+  <tr><td>|T| is `i32`, `u32`, or `f32`
+      <td>*UNSUFFIXED_NUMERIC_LITERAL*: |T|
+      <td>General numeric value.
+</table>
+
+This is a parameterized type rule with three overloads, where `T` is `i32`, `u32` and `f32`.
+Since `7` falls in the required range,
+all overloads apply, and we can derive three type assertions:
+`7: i32`, `7: u32`, and `7: f32`.
+
+Similarly, for the expression `7 + 8`,
+the addition type rules have three overloads that apply:
+
+<table class='data'>
+  <thead>
+    <tr><th>Precondition<th>Conclusion
+  </thead>
 
-A WGSL source program is <dfn noexport>well-typed</dfn> when:
-* The static type can be determined for each expression in the program by applying the type rules, and
-* The type requirements for each statement are satisfied.
+  <tr algorithm="integer addition example">
+    <td>|e1| : |T|<br>|e2| : |T|<br>|T| is `i32` or `u32`
+    <td>|e1| `+` |e2| : |T|
+  <tr algorithm="floating point addition example">
+    <td>|e1| : |T|<br>|e2| : |T|<br>|T| is `f32`
+    <td>|e1| `+` |e2| : |T|
+</table>
 
+These rules' conclusions match a valid parse of the expression `7 + 8`
+where |e1| is `7` and |e2| is `8`.
+If we substitute `i32` for `T` in the first rule,
+the preconditions `7: i32` and `8: i32` can be derived for `7` and `8`,
+so we have derived the assertion `7 + 8: i32`.
+Similarly, we can also derive `7 + 8: u32`.
+Using the second rule, we can further derive `7 + 8: f32`.
+
+However, the expression `7 + 8 + 9.0` is not so flexible.
```

Comment:
I agree with this example, but I don't understand how to apply the rules to get the result. For simplicity, let's parenthesize the expression: `(7 + 8) + 9.0`.  `(7 + 8)` is a type error according to this PR. So is the outer expression analyzed first and its result applied to the subexpression? Sometimes the other way around (e.g. `7 + (8 + 9.0)`)?

The type checking rules currently require subexpressions to be resolved first. Your proposal seems like a constraint solver that must look at the whole expression tree at once. I think that is the major difference from #2227. When constant expressions are fully added David's proposal still works upwards from subexpressions because those subexpression can remain as abstract integers or floats.

Reply:
Right. 

My key concern about @alan-baker's original #2221 was that it was phrased as a top-down algorithm.  I wanted a more bottom-up treatment, because it is can be more explicit how wide a "phrase" to consider when making inferencing decisions.
That's why I started on #2227, trying to achieve the same end but from bottom-up.  I give great credit for Alan's PR for giving very good motivating examples along the lines of the (7+8) + 9.0 case.

I agree with Alan on a key point here, that phrases like   (7+8) have more than one concrete overload and so `let x = 7+8;` is an ill-typed program. But it shouldn't be.  We should prefer the   i32+i32 overload over the u32+u32 overload.

The challenge extends in along another axis when we add other width types.  When we add 16bit integer types, we want these to "work":

       let a: i16 = 7 + 8;   // Should be the same as let a:i16 = i16(15);
       let b = min( i16(7), 8) ; // Should be same as   min(i16(7),i16(8))